### PR TITLE
Add Authorization header authentication

### PR DESCRIPTION
**Pull request by mahmoudhossam**
Based on the discussion in #8, I decided to implement the changes suggested by @NovemberOscar to get them in a PyPI release.

### CODE REVIEW

**Comment by mahmoudhossam** [comment]
Changes were corrupted while merging with the other pull request, will open another PR with the correct changes, closing...

**Comment by NovemberOscar** [comment]
I've just closed #8.

**Comment by mahmoudhossam** [comment]
@NovemberOscar I think it's ready now, feel free to review.

File: flask_graphql_auth/decorators.py [hunk 1]
```diff
@@ -79,6 +79,30 @@ def verify_refresh_jwt_in_argument(token):
     ctx_stack.top.jwt = jwt_data
 
 
+def _extract_token_value(request_params, request_headers, token_name="token"):
+    """
+    Extract token value from the request's params by specific name.
```
Comment (NovemberOscar) [inline flask_graphql_auth/decorators.py]: Does "request param" means query parameter in URL?

File: flask_graphql_auth/decorators.py [hunk 2]
```diff
@@ -79,6 +79,30 @@ def verify_refresh_jwt_in_argument(token):
...
+    :param request_params: Request params as dict
+    :param request_headers: Request headers as dict
+    :param token_name: The name of the JWT token param
+    :return: Token value as a string (empty string if token is not found)
+    """
+    if token_name in request_params:
+        return request_params.pop(token_name)
+    else:
+        authorization_header = request_headers.get("Authorization", "")
+        if authorization_header.lower().startswith("bearer"):
```
Comment (NovemberOscar) [inline flask_graphql_auth/decorators.py]: Please use app configuration to get token location and token prefix, not hard-coded value

File: flask_graphql_auth/decorators.py [hunk 3]
```diff
@@ -100,6 +124,30 @@ def wrapper(*args, **kwargs):
...
+    """
+    A decorator to protect a query resolver.
+
+    If you decorate an resolver with this, it will ensure that the requester
+    has a valid access token before allowing the resolver to be called. This
+    does not check the freshness of the access token.
+    """
+    @wraps(fn)
+    def wrapper(*args, **kwargs):
+        token = _extract_token_value(
```
Comment (NovemberOscar) [inline flask_graphql_auth/decorators.py]: I think "*_header_jwt_required" named decorators should get token from only headers, not in request parameters.

File: tests/test_decorators.py [hunk 4]
```diff
@@ -60,3 +61,28 @@ def test_mutation_refresh_jwt_token_required(flask_app: Flask):
         assert get_jwt_data(response['refresh']["newToken"], "access")["identity"] == "username"
 
 
+def test_extract_token_value_from_params(flask_app):
+    token_name = "token"
+    token_value = "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9"
+    request_params = {token_name: token_value}
+    assert _extract_token_value(request_params, token_name) == token_value
+
```
Comment (NovemberOscar) [inline tests/test_decorators.py]: Please add decorator tests

Comment (mahmoudhossam) [reply tests/test_decorators.py]: I can't seem to be able to construct a query that works with the existing schema without the protected part.

Can you show me an example?

Comment (NovemberOscar) [reply tests/test_decorators.py]: Ok, I'll modify schema and add tests after merging this.

File: flask_graphql_auth/decorators.py [hunk 5]
```diff
@@ -79,6 +79,25 @@ def verify_refresh_jwt_in_argument(token):
     ctx_stack.top.jwt = jwt_data
 
 
+def _extract_header_token_value(current_app, request_headers):
```
Comment (NovemberOscar) [inline flask_graphql_auth/decorators.py]: please remove `current_app` param.

File: tests/test_decorators.py [hunk 6]
```diff
@@ -60,3 +61,13 @@ def test_mutation_refresh_jwt_token_required(flask_app: Flask):
         assert get_jwt_data(response['refresh']["newToken"], "access")["identity"] == "username"
 
 
+def test_extract_header_token_value_from_authorization_header(flask_app):
+    token_value = "eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9"
+    headers = {
+        "Authorization": "Bearer {}".format(token_value)
+    }
+    assert _extract_header_token_value(flask_app, headers) == token_value
```
Comment (NovemberOscar) [inline tests/test_decorators.py]: Test with flask_app.app_context()

ref: http://flask.pocoo.org/docs/1.0/appcontext/#manually-push-a-context

**Comment by mahmoudhossam** [comment]
@NovemberOscar removed the app parameter and used the app context in the tests.

**Comment by NovemberOscar** [comment]
This change will be released soon :)

Thanks for your hard work!

**Pull request merged**